universum/modules/artifact_collector.py

```diff
@@ -128,48 +139,61 @@ def preprocess_artifact_list(self, artifact_list, ignore_already_existing=False)
-            if os.path.exists(path_to_check1) or os.path.exists(path_to_check2):
-                text = f"Build artifact '{os.path.basename(item['path'])}' already present in artifact directory."
-                text += "\nPossible reason of this error: previous build results in working directory"
-                raise CriticalCiException(text)
-
-            dir_list.add(item["path"])
         new_artifact_list = list(dir_list)
         new_artifact_list.sort(key=len, reverse=True)
         return new_artifact_list
 
+
+    def preprocess_artifact(self, item, ignore_already_existing):
+        # Check existence in place: wildcards applied
+        matches = glob2.glob(item["path"])
+        if matches:
+            if item["clean"]:
+                for matching_path in matches:
+                    try:
+                        os.remove(matching_path)  # TODO: use shutil by default
+                    except OSError as e:
+                        if "Is a directory" not in e.strerror:
+                            raise
+                        shutil.rmtree(matching_path)
+                    self.out.log(f"Cleaned up '{matching_path}'")
+            elif not ignore_already_existing:
+                text = "Build artifacts, such as"
+                for matching_path in matches:
+                    text += f"\n * '{os.path.basename(matching_path)}'"
+                text += f"\nalready exist in '{os.path.dirname(item['path'])}' directory."
+                text += "\nPossible reason of this error: previous build results in working directory"
+                raise CriticalCiException(text)
+
+        # Check existence in 'artifacts' directory: wildcards NOT applied
+        path_to_check1 = os.path.join(self.artifact_dir, os.path.basename(item["path"]))
+        path_to_check2 = os.path.join(path_to_check1 + ".zip")
+        if os.path.exists(path_to_check1) or os.path.exists(path_to_check2):
+            text = f"Build artifact '{os.path.basename(item['path'])}' already present in artifact directory."
+            text += "\nPossible reason of this error: previous build results in working directory"
+            raise CriticalCiException(text)
+
+
     @make_block("Preprocessing artifact lists")
     def set_and_clean_artifacts(self, project_configs: Configuration, ignore_existing_artifacts: bool = False) -> None:
         self.html_output.artifact_dir_ready = True
         artifact_list = []
         report_artifact_list = []
         for configuration in project_configs.all():
             if configuration.artifacts:
-                path = utils.parse_path(configuration.artifacts, self.settings.project_root)
-                artifact_list.append(dict(path=path, clean=configuration.artifact_prebuild_clean))
+                artifact_list.append(self.get_config_artifact(configuration))
+            if configuration.is_conditional:
+                artifacts_info = self.get_conditional_step_branches_artifacts(configuration)
```

Comment:
I think you could implement the artifact support by adding artifacts of both branches to the list. All other code should not depend on the steps structure and execution status when dealing with artifacts.
Fortunately or unfortunately this is how artifacts are currently processed (if I am not mistaken).